pettingzoo/sisl/waterworld/waterworld_base.py

```diff
@@ -327,21 +334,55 @@ def _extract_speed_features(self, object_velocities, object_sensorvals, sensed_m
     def collision_handling_subroutine(self, rewards, is_last):
         # Stop pursuers upon hitting a wall
         for pursuer in self._pursuers:
-            clipped_coord = np.clip(pursuer.position, 0, 1)
-            clipped_velocity = pursuer.velocity
-            # If x or y position gets clipped, set x or y velocity to 0 respectively
-            clipped_velocity[pursuer.position != clipped_coord] = 0
-            # Save clipped velocity and position
-            pursuer.set_velocity(clipped_velocity)
-            pursuer.set_position(clipped_coord)
+
+            # here we are trying to clip based on a circle, not a square
+            # the problem amounts to solving a quadratic equation for euclidean distance
+            # given a position and velocity vector, we want to "go back in time" so that
+            # the object is just inside the circle
+            distance = abs(pursuer.position - self.initial_obstacle_coord) + pursuer._radius
+            if (distance[0] ** 2 + distance[1] ** 2 > 0.25):
+                # quadratic equation coefficients for euclidean distance
+                coeff = [pursuer.velocity[0] ** 2 + pursuer.velocity[1] ** 2,
+                         -2 * pursuer.velocity[0] * distance[0] - 2 * pursuer.velocity[1] * distance[1],
+                         distance[0] ** 2 + distance[1] ** 2 - 0.25]
+                ans = np.roots(coeff)
+                # we find the smallest positive t that confines in the circle
+                t = min(abs(ans))
+                pursuer.set_position(pursuer.position - t * pursuer.velocity)
+                pursuer.set_velocity(np.array([0, 0]))
```

Comment:
It would probably be easier and more computationally efficient if the agent velocity and position was converted to polar form (here we actually only need the radius component of both position and velocity):
```
polar_dis = np.linalg.norm(pursuer.position - 0.5)
polar_vel = np.linalg.norm(pursuer.velocity)
```
Then find out what ratio of time is required for `polar_dis` to be at the edge of the circle:
```
t = (circle_radius - pursuer_position) / polar_vel
```

Reply:
Can probably calculate the remaining time by simply recycling the above calculation for `L4`, and then just doing:
```py
t = L4 / np.linalg.norm(pursuer.velocity)
```